File: backend/models/CodingReport.js
const mongoose = require('mongoose');

const testCaseResultSchema = new mongoose.Schema({
  input: String,
  expected: String,
  actual: String,
  status: { type: String, enum: ['Pass', 'Fail'] }
}, { _id: false });

const codingReportSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  problemId: {
    type: Number,
    required: true
  },
  problemTitle: {
    type: String,
    required: true
  },
  language: {
    type: String,
    required: true
  },
  code: {
    type: String,
    required: true
  },
  // Overall result
  score: {
    type: Number,  // 0 – 10
    default: 0
  },
  verdict: {
    type: String,  // e.g. 'Accepted', 'Wrong Answer', 'Compile Error'
    default: ''
  },
  // Test case summary
  testCasesPassed: {
    type: Number,
    default: 0
  },
  testCasesTotal: {
    type: Number,
    default: 0
  },
  testCaseResults: [testCaseResultSchema],
  // GROQ evaluation fields
  feedback: {
    type: String,
    default: ''
  },
  hints: {
    type: String,
    default: ''
  },
  timeComplexity: {
    type: String,
    default: ''
  },
  // Extended evaluation fields (populated when available)
  spaceComplexity: {
    type: String,
    default: ''
  },
  codeQuality: {
    type: Number,  // 0 – 10
    default: null
  },
  readability: {
    type: Number,  // 0 – 10
    default: null
  },
  optimization: {
    type: Number,  // 0 – 10
    default: null
  },
  edgeCases: {
    type: Number,  // 0 – 10
    default: null
  },
  bestPractices: {
    type: Number,  // 0 – 10
    default: null
  },
  strengths: {
    type: String,
    default: ''
  },
  weaknesses: {
    type: String,
    default: ''
  },
  recommendations: {
    type: String,
    default: ''
  }
}, { timestamps: true });

module.exports = mongoose.model(
  'CodingReport',
  codingReportSchema,
  'coding_evaluation_reports'
);

File: backend/models/coding_report.py
"""models/coding_report.py — CodingReport table (replaces Mongoose CodingReport.js)"""
from sqlalchemy import Column, Integer, String, Float, Boolean, DateTime, ForeignKey, Text, JSON
from sqlalchemy.sql import func
from database import Base


class CodingReport(Base):
    __tablename__ = "coding_reports"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    problem_id = Column(Integer, nullable=False)
    problem_title = Column(String(500), nullable=False)
    language = Column(String(100), nullable=False)
    code = Column(Text, nullable=False)
    score = Column(Float, default=0)        # 0–10
    verdict = Column(String(255), default="")
    test_cases_passed = Column(Integer, default=0)
    test_cases_total = Column(Integer, default=0)
    test_case_results = Column(JSON, default=list)  # [{input, expected, actual, status}]
    feedback = Column(Text, default="")
    hints = Column(Text, default="")
    time_complexity = Column(String(100), default="")
    space_complexity = Column(String(100), default="")
    code_quality = Column(Float, nullable=True)
    readability = Column(Float, nullable=True)
    optimization = Column(Float, nullable=True)
    edge_cases = Column(Float, nullable=True)
    best_practices = Column(Float, nullable=True)
    strengths = Column(Text, default="")
    weaknesses = Column(Text, default="")
    recommendations = Column(Text, default="")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
